src/Arcus.Messaging.Pumps.ServiceBus/Resiliency/CircuitBreakerServiceBusMessageHandler.cs

```diff
@@ -0,0 +1,108 @@
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CircuitBreakerServiceBusMessageHandler{TMessage}" /> class.
+        /// </summary>
+        /// <param name="circuitBreaker">The circuit breaker that controls the activation of the message pump.</param>
+        /// <param name="logger">The logger to write diagnostic messages during the processing of the message.</param>
+        /// <exception cref="ArgumentNullException">Thrown when the <paramref name="circuitBreaker"/> or <paramref name="logger"/> is <c>null</c>.</exception>
+        protected CircuitBreakerServiceBusMessageHandler(
+            IMessagePumpCircuitBreaker circuitBreaker,
+            ILogger<CircuitBreakerServiceBusMessageHandler<TMessage>> logger) 
+            : base(logger)
+        {
+            CircuitBreaker = circuitBreaker;
+        }
+
+        /// <summary>
+        /// Gets the circuit breaker that controls the activation of the message pump.
+        /// </summary>
+        protected IMessagePumpCircuitBreaker CircuitBreaker { get; }
+
+        /// <summary>
+        /// Process a new message that was received.
+        /// </summary>
+        /// <param name="message">The message that was received.</param>
+        /// <param name="messageContext">The context providing more information concerning the processing.</param>
+        /// <param name="correlationInfo">The information concerning correlation of telemetry and processes by using a variety of unique identifiers.</param>
+        /// <param name="cancellationToken">The token to cancel the processing.</param>
+        /// <exception cref="ArgumentNullException">
+        ///     Thrown when the <paramref name="message"/>, <paramref name="messageContext"/>, or the <paramref name="correlationInfo"/> is <c>null</c>.
+        /// </exception>
+        public override async Task ProcessMessageAsync(
+            TMessage message,
+            AzureServiceBusMessageContext messageContext,
+            MessageCorrelationInfo correlationInfo,
+            CancellationToken cancellationToken)
+        {
+            var options = new MessagePumpCircuitBreakerOptions();
+            MessageProcessingResult result = await TryProcessMessageAsync(message, messageContext, correlationInfo, options, cancellationToken);
+
+            if (!result.IsSuccessful)
+            {
+                await CircuitBreaker.PauseMessageProcessingAsync(messageContext.JobId, opt =>
+                {
+                    opt.MessageIntervalDuringRecovery = options.MessageIntervalDuringRecovery;
+                    opt.MessageRecoveryPeriod = options.MessageRecoveryPeriod;
+                });
+                await AbandonMessageAsync();
+                throw result.ProcessingException;
+            }
+
+            await CircuitBreaker.ResumeMessageProcessingAsync(messageContext.JobId);
```

Comment:
We should not always just call ResumeMessageProcessing I think.
This should only be called when the circuit-breaker is in the half-open state.
(Suppose the circuitbreaker is just triggered, but you still have the remainder of the messages that have been retrieved by the pump that must be processed.  You don't want to close the circuitbreaker again because one of those messages can be processed.).

Reply:
This is no longer relevant.  The application developer is not responsible for calling resume himself with the changes we've made now.  
The MessagePump will automatically transition into a half-open and later in a closed state.